--- linux-kukui-gud/arch/arm64/boot/dts/mediatek/mt8183.dtsi	2021-07-13 17:01:43.516154051 +0200
+++ linux-kukui/arch/arm64/boot/dts/mediatek/mt8183.dtsi	2021-07-13 17:39:36.518680936 +0200
@@ -1732,6 +1732,8 @@
 			reg = <0 0x14016000 0 0x1000>;
 			interrupts = <GIC_SPI 217 IRQ_TYPE_LEVEL_LOW>;
 			power-domains = <&spm MT8183_POWER_DOMAIN_DISP>;
+			mediatek,gce-events = <CMDQ_EVENT_MUTEX_STREAM_DONE0>,
+					      <CMDQ_EVENT_MUTEX_STREAM_DONE1>;
 		};
 
 		larb0: larb@14017000 {
