diff --git a/drivers/gpu/drm/panel/panel-boe-tv101wum-nl6.c b/drivers/gpu/drm/panel/panel-boe-tv101wum-nl6.c
index db9d0b86d..9cee37f27 100644
--- a/drivers/gpu/drm/panel/panel-boe-tv101wum-nl6.c
+++ b/drivers/gpu/drm/panel/panel-boe-tv101wum-nl6.c
@@ -561,14 +561,15 @@ static int boe_panel_prepare(struct drm_panel *panel)
 	ret = boe_panel_init_dcs_cmd(boe);
 	if (ret < 0) {
 		dev_err(panel->dev, "failed to init panel: %d\n", ret);
-		goto poweroff;
+		goto poweroffavee;
 	}
 
+	msleep(130);
 	boe->prepared = true;
 
 	return 0;
 
-poweroff:
+poweroffavee:
 	regulator_disable(boe->avee);
 poweroffavdd:
 	regulator_disable(boe->avdd);
@@ -576,16 +577,10 @@ static int boe_panel_prepare(struct drm_panel *panel)
 	usleep_range(5000, 7000);
 	regulator_disable(boe->pp1800);
 	gpiod_set_value(boe->enable_gpio, 0);
+	boe->prepared = false;
 
 	return ret;
 }
-
-static int boe_panel_enable(struct drm_panel *panel)
-{
-	msleep(130);
-	return 0;
-}
-
 static const struct drm_display_mode boe_tv101wum_nl6_default_mode = {
 	.clock = 159425,
 	.hdisplay = 1200,
@@ -742,7 +737,8 @@ static int boe_panel_get_modes(struct drm_panel *panel,
 	connector->display_info.width_mm = boe->desc->size.width_mm;
 	connector->display_info.height_mm = boe->desc->size.height_mm;
 	connector->display_info.bpc = boe->desc->bpc;
-	drm_connector_set_panel_orientation(connector, boe->orientation);
+	drm_connector_set_panel_orientation(connector, DRM_MODE_PANEL_ORIENTATION_LEFT_UP);
+	//drm_connector_set_panel_orientation(connector, boe->orientation);
 
 	return 1;
 }
@@ -750,7 +746,6 @@ static int boe_panel_get_modes(struct drm_panel *panel,
 static const struct drm_panel_funcs boe_panel_funcs = {
 	.unprepare = boe_panel_unprepare,
 	.prepare = boe_panel_prepare,
-	.enable = boe_panel_enable,
 	.get_modes = boe_panel_get_modes,
 };
 
