diff -Nur linux-kukui/drivers/opp/core.c linux-kukui-gud/drivers/opp/core.c
--- linux-kukui/drivers/opp/core.c	2021-06-21 00:03:15.000000000 +0200
+++ linux-kukui-gud/drivers/opp/core.c	2021-06-23 19:43:59.735379192 +0200
@@ -756,18 +756,15 @@
 	struct regulator *reg = opp_table->regulators[0];
 	struct dev_pm_opp *old_opp = opp_table->current_opp;
 	int ret;
-
-	/* This function only supports single regulator per device */
-	if (WARN_ON(opp_table->regulator_count > 1)) {
-		dev_err(dev, "multiple regulators are not supported\n");
-		return -EINVAL;
-	}
+	unsigned i;
 
 	/* Scaling up? Scale voltage before frequency */
 	if (!scaling_down) {
-		ret = _set_opp_voltage(dev, reg, opp->supplies);
-		if (ret)
-			goto restore_voltage;
+		for (i = 0; i < opp_table->regulator_count; i++) {
+			ret = _set_opp_voltage(dev, opp_table->regulators[i], &opp->supplies[i]);
+			if (ret)
+				goto restore_voltage;
+		}
 	}
 
 	/* Change frequency */
@@ -777,9 +774,11 @@
 
 	/* Scaling down? Scale voltage after frequency */
 	if (scaling_down) {
-		ret = _set_opp_voltage(dev, reg, opp->supplies);
-		if (ret)
-			goto restore_freq;
+		for (i = 0; i < opp_table->regulator_count; i++) {
+			ret = _set_opp_voltage(dev, opp_table->regulators[i], &opp->supplies[i]);
+			if (ret)
+				goto restore_freq;
+		}
 	}
 
 	/*
