--- linux-kukui-new/arch/arm64/boot/dts/mediatek/mt6358.dtsi	2021-06-30 16:08:03.382271887 +0200
+++ linux-kukui/arch/arm64/boot/dts/mediatek/mt6358.dtsi	2021-06-30 16:10:18.072445277 +0200
@@ -209,6 +209,7 @@
 
 			mt6358_vsram_gpu_reg: ldo_vsram_gpu {
 				regulator-name = "vsram_gpu";
+				regulator-always-on;
 				regulator-min-microvolt = <500000>;
 				regulator-max-microvolt = <1293750>;
 				regulator-ramp-delay = <6250>;
